lib/collection/src/hash_ring.rs

```diff
@@ -44,24 +44,24 @@ impl<T: Hash + Copy + PartialEq> HashRingRouter<T> {
         ring
     }
 
-    pub fn is_empty(&self) -> bool {
-        match self {
-            Self::Single(ring) => ring.is_empty(),
-            Self::Resharding { old, new } => old.is_empty() && new.is_empty(),
-        }
-    }
-
-    pub fn is_resharding(&self) -> bool {
-        matches!(self, Self::Resharding { .. })
-    }
-
-    pub fn add(&mut self, shard: T) {
+    pub fn add(&mut self, shard: T) -> bool {
         match self {
             Self::Single(ring) => ring.add(shard),
+
             Self::Resharding { old, new } => {
-                if new.get(&shard).is_none() {
+                // When resharding is in progress:
+                // - either `new` hashring contains a shard, that is not in `old` (when resharding *up*)
+                // - or `old` contains a shard, that is not in `new` (when resharding *down*)
+                //
+                // This check ensures, that we don't accidentally break this invariant when adding
+                // nodes to `Resharding` hashring.
+
+                if !old.contains(&shard) && !new.contains(&shard) {
                     old.add(shard);
                     new.add(shard);
+                    true
+                } else {
+                    false
                 }
             }
         }
```

Comment:
I don't understand how this works.

Why wouldn't we only add in the new hash ring here?

Reply:
We decided to simply forbid adding shards into `HashRingRouter::Resharding` 🤷‍♀️